### PR TITLE
Multiple `wibox.layout` enhancements

### PR DESCRIPTION
**Pull request by Elv13**
Multiple `wibox.layout` enhancements required for the client layout implementation along with some `awful.util.table` methods to reduce boilerplate code.

### CODE REVIEW

**Comment by psychon** [comment]
To make sure they don't get lost: My comments are at https://github.com/Elv13/awesome-1/commit/78c70295fe3f1b85acf49e2840a6c351a99e1c5c plus the parent commits

**Comment by Elv13** [comment]
Updated, I removed the `ratio` and `stack` from this PR. There is enough work getting this one in first, lets simplify my life a little

**Comment by Elv13** [comment]
Updated

**Comment by Elv13** [comment]
Feel free to review this one, but I will make another one where there is method with `index` and one with `widget` to add some consistency. So `:remove(index)` and `:remove_widgets(..., recursive)`.

File: lib/wibox/widget/base.lua [hunk 1]
```diff
@@ -482,6 +490,30 @@ function base.check_widget(widget)
...
 end
 
+--- Get a widex index
+-- @param widget The widget to look for
+-- @param[opt] recursive Also check sub-widgets
+-- @param[opt] ... Aditional widgets to add at the end of the "path"
+-- @return The index
+-- @return The parent layout
+-- @return The path between "self" and "widget"
+function base:index(widget, recursive, ...)
```
Comment (Elv13) [inline lib/wibox/widget/base.lua]: Oops, this should be base.widget

File: lib/wibox/layout/flex.lua [hunk 2]
```diff
@@ -6,16 +6,55 @@
...
+--   a stack overflow
+-- @param[opt] recursive Also add all widgets of childrens
+-- @return a list of all widgets
+-- @name get_widgets
+-- @class function
+
+--- Get the number of children element
+-- @param layout The layout you are modifying.
+-- @return The number of children element
+-- @name get_children_count
```
Comment (Elv13) [inline lib/wibox/layout/flex.lua]: Remove this, replace with get_children

File: lib/wibox/layout/flex.lua [hunk 3]
```diff
@@ -6,16 +6,55 @@
...
+-- @name layout
+-- @class function
+
+--- Get all children of this layout
+-- @param layout The layout you are modifying.
+-- @warning If the widget contain itself and recursive is true, this will cause
+--   a stack overflow
+-- @param[opt] recursive Also add all widgets of childrens
+-- @return a list of all widgets
+-- @name get_widgets
```
Comment (Elv13) [inline lib/wibox/layout/flex.lua]: Remove this

File: lib/wibox/widget/base.lua [hunk 4]
```diff
@@ -482,6 +490,30 @@ function base.check_widget(widget)
...
+    local widgets = self.get_widgets and self:get_widgets() or {}
```
Comment (Elv13) [inline lib/wibox/widget/base.lua]: get_children

File: lib/wibox/layout/fixed.lua [hunk 5]
```diff
@@ -62,6 +63,138 @@ function fixed:add(...)
...
+                self_changed = true
+            end
+        else
+            ret = false
+        end
+
+    end
+
+    if self_changed then
+        l:emit_signal("widget::layout_changed")
```
Comment (Elv13) [inline lib/wibox/layout/fixed.lua]: s/l/self/

File: lib/wibox/layout/fixed.lua [hunk 6]
```diff
@@ -62,6 +63,172 @@ function fixed:add(...)
...
+    local ret = true
+    for k, rem_widget in ipairs(args) do
+        if recursive and k == #args then break end
+
+        local idx, l = self:index(rem_widget, recursive)
+
+        if idx and l then
+            -- In case :remove() is overloaded, make sure its logic is executed
+            if l ~= self and l.remove then
+                l:remove(rem_widget, false)
```
Comment (Elv13) [inline lib/wibox/layout/fixed.lua]: s/rem_widget/idx

**Comment by Elv13** [comment]
Any comments? Is this one good enough to be merged? I need it so I can update some of the other PRs (as they would depend on this one after the changes)

**Comment by psychon** [comment]
You win. The recursive stuff is now almost simple enough that I can live with it. However (as already written on that comment), I'm still complaining about `:get_children(true)` and want it split up. Sorry.

Besides that, this looks good to me and has mostly minor issues.

Basing `layout.flex` on `layout.fixed` still bothers me a bit. This commit already forgets to update the docs there... Perhaps this should just say something like `This layout inherits from the fixed layout and thus all methods there are also available here. @see wibox.layout.fixed` in the general description at the beginning of the file (or something like that)?

**Comment by Elv13** [comment]
Thanks for the review, I will fix that tomorrow.

**Comment by Elv13** [comment]
Updated, I still need to rebase the other 6 PR (this is PR 5/10 for the dynamic layout + the declarative widget PR) on top of this one, so it is still untested (this revision) beside being able to load my config and the default one.

**Comment by Elv13** [comment]
Ok, I tested it, work as expected (I hope I won't have to rebase all these patches again, this take so long...)

**Comment by Elv13** [comment]
All fixed (btw, you should really add comments in the `Files changed` section, this way they would not get lost every single time. Sorry for the untested `:swap()`, apparently all my code used the recursive one

File: lib/wibox/layout/align.lua [hunk 7]
```diff
@@ -164,6 +165,12 @@ function align:set_third(widget)
     self:emit_signal("widget::layout_changed")
 end
 
+--- Get all children of this layout
+-- @treturn table a list of all widgets
+function align:get_children()
+    return {self.first, self.second, self.third}
```
Comment (psychon) [inline lib/wibox/layout/align.lua]: I think the `from_sparse` here is still needed. Sorry!
(If `first` and `second` is nil.... oh and the recursive function will do `(nil):get_children()`)

File: lib/wibox/layout/constraint.lua [hunk 8]
```diff
@@ -8,6 +8,7 @@
 local pairs = pairs
 local type = type
 local setmetatable = setmetatable
+local util = require("awful.util")
```
Comment (psychon) [inline lib/wibox/layout/constraint.lua]: Unused import

File: lib/wibox/layout/constraint.lua [hunk 9]
```diff
@@ -44,6 +45,15 @@ function constraint:set_widget(widget)
     self:emit_signal("widget::layout_changed")
 end
 
+--- Get the number of children element
+-- @warning If the widget contain itself and recursive is true, this will cause
+--   a stack overflow
+-- @tparam[opt=false] boolean recursive Also add all widgets of childrens
+-- @treturn table The children
+function constraint:get_children(recursive)
```
Comment (psychon) [inline lib/wibox/layout/constraint.lua]: Unused argument

File: lib/wibox/layout/margin.lua [hunk 10]
```diff
@@ -9,6 +9,7 @@ local pairs = pairs
 local type = type
 local setmetatable = setmetatable
 local base = require("wibox.widget.base")
+local util = require("awful.util")
```
Comment (psychon) [inline lib/wibox/layout/margin.lua]: unused import and the `recursive` argument below isn't used either (I'll stop making this comment on all instances of this)

File: lib/wibox/layout/fixed.lua [hunk 11]
```diff
@@ -62,6 +63,140 @@ function fixed:add(...)
...
+
+    return true
+end
+
+--- Insert a new widget in the layout at position `index`
+-- @tparam number index The position
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+function fixed:insert(index, widget)
+    if index and index > 0 and index > #self.widgets + 1 then return false end
```
Comment (psychon) [inline lib/wibox/layout/fixed.lua]: `if not index or index < 1 or index > #self.widgets + 1 then return false end`

**Comment by Elv13** [comment]
fixed, again (shame on me for the insert one)

**Comment by psychon** [comment]
Thanks!

**Comment by Elv13** [comment]
Can I merge it?

**Comment by psychon** [comment]
Fine by me. That's what the milestone means, after all. :-) (Sorry if I wasn't clear on that)

**Pull request merged**